drivers/motor_driver/motor_driver.c

```diff
@@ -0,0 +1,209 @@
+    motor_direction_t direction, uint16_t pwm_duty_cycle)
+{
+    assert(motor_driver < MOTOR_DRIVER_NUMOF);
+
+    const motor_driver_config_t *motor_driver_conf =
+        &motor_driver_config[motor_driver];
+
+    assert(motor_id < motor_driver_conf->nb_motors);
+
+    const motor_t *dev = &motor_driver_conf->motors[motor_id];
+
+    direction = direction ^ dev->gpio_dir_reverse;
+
+    /* Two direction GPIO, handling brake */
+    if (motor_driver_conf->mode == MOTOR_DRIVER_2_DIRS)
+    {
+        if ((dev->gpio_dir0 == GPIO_UNDEF) || \
+            (dev->gpio_dir1_or_brake == GPIO_UNDEF))
+        {
+            errno = ENODEV;
+            goto motor_set_err;
+        }
+        switch (direction)
+        {
+            case MOTOR_CW:
+            case MOTOR_CCW:
+                /* Direction */
+                gpio_write(dev->gpio_dir0, direction);
+                gpio_write(dev->gpio_dir1_or_brake, direction ^ 0x1);
+                break;
+            case MOTOR_BRAKE:
+            default:
+                /* Brake */
+                gpio_write(dev->gpio_dir0,
+                    motor_driver_conf->mode_brake);
+                gpio_write(dev->gpio_dir1_or_brake,
+                    motor_driver_conf->mode_brake);
+                pwm_duty_cycle = 0;
+                break;
+        }
+    }
+    /* Single direction GPIO */
+    else if (motor_driver_conf->mode == MOTOR_DRIVER_1_DIR)
+    {
+        if (dev->gpio_dir0 == GPIO_UNDEF)
+        {
+            errno = ENODEV;
+            goto motor_set_err;
+        }
+        switch (direction)
+        {
+            case MOTOR_CW:
+            case MOTOR_CCW:
+                /* Direction */
+                gpio_write(dev->gpio_dir0, direction);
+                break;
+
+            case MOTOR_BRAKE:
+            default:
+                pwm_duty_cycle = 0;
+                break;
+        }
+    }
+    /* Single direction GPIO and brake GPIO */
+    else if (motor_driver_conf->mode == MOTOR_DRIVER_1_DIR_BRAKE)
+    {
+        if ((dev->gpio_dir0 == GPIO_UNDEF) || \
+            (dev->gpio_dir1_or_brake == GPIO_UNDEF))
+        {
+            errno = ENODEV;
+            goto motor_set_err;
+        }
+        switch (direction)
+        {
+            case MOTOR_CW:
+            case MOTOR_CCW:
+                /* Direction */
+                gpio_write(dev->gpio_dir0, direction);
+                /* No brake */
+                gpio_write(dev->gpio_dir1_or_brake, dev->gpio_brake_invert);
+                break;
+
+            case MOTOR_BRAKE:
+            default:
+                /* No brake */
+                gpio_write(dev->gpio_dir1_or_brake, 1 ^ dev->gpio_brake_invert);
+                pwm_duty_cycle = 0;
+                break;
+        }
+    }
+    else
+    {
+        errno = EINVAL;
+        goto motor_set_err;
+    }
+
+    pwm_set(motor_driver_conf->pwm_dev, dev->pwm_channel, pwm_duty_cycle);
```

Comment:
You are setting the direction before while the old PWM value is active, that is also a possible glitch, and, again, the CPU could be interrupted between that point and this one.

Reply:
Same than for GPIO calls, I'll check that two.